src/service/feedback.tsx

```diff
@@ -0,0 +1,166 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import React from 'react';
+
+import { Dialog, Notification } from '@jupyterlab/apputils';
+import { ReactWidget } from '@jupyterlab/ui-components';
+
+import { postFeedback } from './api';
+import { getCurrentModel } from './modelHandler';
+import { lastPrompt } from '../QiskitCompletionProvider';
+import { feedbackIcon } from '../utils/icons';
+import { IFeedbackForm, IFeedbackResponse } from '../utils/schema';
+
+const toBool = (bool?: string): boolean | undefined => {
+  switch (bool) {
+    case 'true':
+      return true;
+    case 'false':
+      return false;
+    default:
+      return undefined;
+  }
+};
+
+export function getFeedbackStatusBarWidget() {
+  const feedbackButton = ReactWidget.create(
+    <div
+      className="jp-qiskit-code-assistant-feedback"
+      title="Give feedback for the Qiskit Code Assistant"
+      onClick={() => getFeedback(false)}
+    >
+      <feedbackIcon.react
+        stylesheet="statusBar"
+        tag="span"
+        top="2px"
+        verticalAlign="middle"
+      />
+    </div>
+  );
+
+  feedbackButton.addClass('jp-mod-highlighted');
+
+  return feedbackButton;
+}
+
+export function getFeedback(prompt: boolean = true) {
+  const model_id = getCurrentModel()?._id;
+  const prompt_id = prompt ? lastPrompt?.prompt_id : undefined;
+  const prompt_input = prompt ? lastPrompt?.input : undefined;
+  const prompt_output = prompt ? lastPrompt?.items[0] : undefined;
+
+  const dialog = new Dialog({
+    title: 'Qiskit Code Assistant Feedback',
+    body: getFeedbackBodyWidget(prompt_output),
+    buttons: [Dialog.cancelButton(), Dialog.okButton({ label: 'Submit' })],
+    focusNodeSelector: '.jp-qiskit-code-assistant-feedback-form textarea'
+  });
+
+  const dialogHandleEvent = dialog.handleEvent;
+  dialog.handleEvent = (event: Event) => {
+    if (
+      event.type === 'keydown' &&
+      (event as KeyboardEvent).code === 'Enter' &&
+      document.activeElement instanceof HTMLTextAreaElement
+    ) {
+      return;
+    }
+    dialogHandleEvent.call(dialog, event);
+  };
+
+  return dialog.launch().then(result => {
+    if (
+      result.button.accept &&
+      (result.value?.positive_feedback || result.value?.comment)
+    ) {
+      postFeedback(
+        model_id,
+        prompt_id,
+        toBool(result.value?.positive_feedback),
+        result.value?.comment,
+        prompt_input,
+        prompt_output
+      ).then((response: IFeedbackResponse) => {
+        Notification.info(`Qiskit Code Assistant:\n${response['message']}`, {
+          autoClose: 5000
+        });
+      });
+    }
+  });
+}
+
+function getFeedbackBodyWidget(
+  prompt?: string
+): Dialog.IBodyWidget<IFeedbackForm> {
+  const bodyWidget = ReactWidget.create(
+    <>
+      {!!prompt && (
+        <>
+          <b>Completion</b>
+          <pre>{prompt.trim()}</pre>
```

Comment:
would this better off as a `textarea` with `readonly` attribute?

Reply:
Just did a bit of research and no, a readonly texture is not considered a good was to show formatted text.

While I was at it I checked to see if there was a simple way I missed to insert a read-only CodeMirrorEditor here like I had originally planned to match the design doc, but I only found two solutions, to use a third party lib or write a bunch of code to allow the extension to use the jupyterlab editor factory to generate it (I gave this a try and it's turned out to be even more difficult that just adding a bunch of code)

Reply:
👍🏽